### PR TITLE
LP-218 API transaction interceptor to get transaction

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link
https://companieshouse.atlassian.net/browse/LP-218

### Change description
TBD

### Work checklist

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change
* [x] Infrastructure change

### Pull request checklist

* [ ] I have added unit tests for new code that I have added
* [ ] I have added/updated functional tests where appropriate
* [ ] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__
### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java [hunk 1]
```diff
@@ -30,10 +34,17 @@ public InterceptorConfig(LoggingInterceptor loggingInterceptor, CustomUserAuthen
      */
     @Override
     public void addInterceptors(@NonNull InterceptorRegistry registry) {
+        registry.addInterceptor(transactionInterceptor())
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java:37]: logging interceptor should always come first so it will log everything after

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/utils/Constants.java [hunk 2]
```diff
@@ -8,5 +8,5 @@ private Constants() { }
     public static final String ERIC_REQUEST_ID_KEY = "X-Request-Id";
 
     // URL path parameters
-    public static final String URL_PARAM_TRANSACTION_ID = "transaction_id";
+    public static final String URL_PARAM_TRANSACTION_ID = "transactionId";
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/utils/Constants.java:11]: does the spec need updating for the url param name change

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java:38]: Reordered the config.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/utils/Constants.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/utils/Constants.java:11]: Yes it does. Updated the spec.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by markpit** [comment]
LGTM, seems like coverage issue is due to pipeline not running tests/report correctly

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.io.IOException;
+
+@Component
+public class ApiClientServiceImpl implements ApiClientService {
+
+
+    @Override
+    public ApiClient getApiClient() {
+        return ApiSdkManager.getSDK();
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:17]: You could add a test for these, just check that you get a class of the correct type back and not null?

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:17]: I have taken a look, I'd have a lot of trouble mocking this, the libraries it calls use a lot of resources.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.api.ApiClient;
+import uk.gov.companieshouse.api.InternalApiClient;
+import uk.gov.companieshouse.api.sdk.ApiClientService;
+import uk.gov.companieshouse.sdk.manager.ApiSdkManager;
+
+import java.io.IOException;
+
+@Component
+public class ApiClientServiceImpl implements ApiClientService {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:12]: Just wondering where this class is being used and how? I can't see any references to it.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java (hunk 3 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:17]: you don't have to mock it, just call getApiClient() and see if you get an ApiClient back.

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:17]: It would make no sense mocking it as it's the only line in this method. If you don't mock it, in a test, does the call return an object or fall over?

(In the ROE API we didn't unit test this class, but that doesn't necessarily mean we shouldn't add a useful test here, if achievable)

**Comment by mwestacott** [comment]
Should this PR also be adding the transaction id to the controller method params and then ensuring the call to the new transaction interceptor is working? I'd have thought so.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java (hunk 3 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:17]: I tried to just call the function but the test fails because it requires environment variables referred to within the library, so it does look like we need to mock something here we cannot just call it.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java (hunk 4 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:12]: This implements an interface so requires all 4 methods to be implemented or the compiler will complain.

You won't see references it is an implementation of a interface referred to in the library. I recommend looking at the TransactionInterceptor class in the library. This library requires an implementation otherwise the app will not start.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+    @Override
+    public ApiClient getApiClient() {
+        return ApiSdkManager.getSDK();
+    }
+
+    @Override
+    public ApiClient getApiClient(String ericPassThroughHeader) throws IOException {
```
[the remark this answers is not in the document]
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:21]: Again it's all about the library, recommend viewing it: 
https://github.com/companieshouse/api-security-java/blob/main/src/main/java/uk/gov/companieshouse/api/interceptor/TransactionInterceptor.java

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java (hunk 4 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:12]: 👍 makes sense now. Thanks.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/client/ApiClientServiceImpl.java:21]: Method names need to match those defined in the interface. Now clear.

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java [hunk 6]
```diff
@@ -31,9 +35,16 @@ public InterceptorConfig(LoggingInterceptor loggingInterceptor, CustomUserAuthen
...
+                .addPathPatterns(FILINGS, TRANSACTIONS);
         registry.addInterceptor(new TokenPermissionsInterceptor())
                 .addPathPatterns(PARTNERSHIP);
         registry.addInterceptor(customUserAuthenticationInterceptor)
                 .addPathPatterns(PARTNERSHIP);
     }
+
+    @Bean
+    public TransactionInterceptor transactionInterceptor() {
+        return new TransactionInterceptor();
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java:48]: Could invoke the constructor that adds a logging name space.
e.g. https://github.com/companieshouse/psc-filing-api/blob/b50e10855c783a7c3de5fdcf398379cd51b4775b/src/main/java/uk/gov/companieshouse/pscfiling/api/config/InterceptorConfig.java#L125

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java (hunk 6 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/config/InterceptorConfig.java]: Added the argument

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 7]
```diff
@@ -35,11 +38,12 @@ public PartnershipController(LimitedPartnershipService limitedPartnershipService
 
     @PostMapping
     public ResponseEntity<Object> createPartnership(
+            @RequestAttribute(TRANSACTION_KEY) Transaction transaction,
```
[the remark this answers is not in the document]
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:41]: It wont't work they refer to different values. I took a look and could not find anywhere else to access the same value.

[the remark this answers is not in the document]
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:41]: 👍 I still wonder why request attribute name is "transaction" and where that's defined... Leaving comment unresolved just in case others know.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:41]: It's based on what we did for ROE, which has TRANSACTION_KEY = "transaction";

[the remark this answers is not in the document]
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:41]: Can we resolve this?

[the remark this answers is not in the document]
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:41]: Yep. Main thing is it works.

**Review by mwestacott — APPROVED** [review]

**Review by markpit — APPROVED** [review]
